File: frontend/src/api/trips.ts
import { apiFetch } from './client'
import type { MessageResponse, TripResponse } from '../types/api'

export function saveTrip(trip: TripResponse) {
  return apiFetch<TripResponse>('/trips', { method: 'POST', body: JSON.stringify(trip) })
}

export function listTrips() {
  return apiFetch<TripResponse[]>('/trips')
}

export function getTrip(placeName: string) {
  return apiFetch<TripResponse>(`/trips/${encodeURIComponent(placeName)}`)
}

export function deleteTrip(placeName: string) {
  return apiFetch<MessageResponse>(`/trips/${encodeURIComponent(placeName)}`, { method: 'DELETE' })
}

export function startPlanning(placeName: string) {
  return apiFetch<MessageResponse>(`/trips/${encodeURIComponent(placeName)}/plan`, { method: 'POST' })
}

File: backend/app/routers/trips.py
from fastapi import APIRouter, HTTPException, status

from app.deps import CurrentUser, Db
from app.schemas.trip import MessageResponse, SaveTripRequest, TripResponse
from app.services import trip_service

router = APIRouter(prefix="/trips", tags=["trips"])


@router.post("", response_model=TripResponse)
async def save_trip(payload: SaveTripRequest, user: CurrentUser, db: Db) -> TripResponse:
    trip_dict = payload.to_legacy_dict()
    status_result = await trip_service.save_trip(db, user.username, trip_dict)
    if status_result == "exists":
        raise HTTPException(status.HTTP_409_CONFLICT, detail="Trip already saved")
    return TripResponse.from_legacy_dict(trip_dict)


@router.get("", response_model=list[TripResponse])
async def list_trips(user: CurrentUser, db: Db) -> list[TripResponse]:
    return await trip_service.get_saved_trips(db, user.username)


@router.get("/{place_name}", response_model=TripResponse)
async def get_trip(place_name: str, user: CurrentUser, db: Db) -> TripResponse:
    trip = await trip_service.get_saved_trip(db, user.username, place_name)
    if trip is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Trip not found")
    return trip


@router.delete("/{place_name}", response_model=MessageResponse)
async def delete_trip(place_name: str, user: CurrentUser, db: Db) -> MessageResponse:
    deleted = await trip_service.delete_saved_trip(db, user.username, place_name)
    if not deleted:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Trip not found")
    return MessageResponse(message="Trip deleted")


@router.post("/{place_name}/plan", response_model=MessageResponse)
async def start_planning(place_name: str, user: CurrentUser, db: Db) -> MessageResponse:
    ok = await trip_service.start_planning(db, user.username, place_name)
    if not ok:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Trip not found in saved trips")
    return MessageResponse(message="Planning started")
